compiler/rustc_middle/src/ty/mod.rs

```diff
@@ -1625,12 +1625,15 @@ impl<'tcx> TyCtxt<'tcx> {
         def_id1: DefId,
         def_id2: DefId,
     ) -> Option<ImplOverlapKind> {
-        let impl1 = self.impl_trait_header(def_id1).unwrap().instantiate_identity();
-        let impl2 = self.impl_trait_header(def_id2).unwrap().instantiate_identity();
+        let impl1 = self.impl_trait_header(def_id1).unwrap();
+        let impl2 = self.impl_trait_header(def_id2).unwrap();
+
+        let trait_ref1 = impl1.trait_ref.instantiate_identity();
+        let trait_ref2 = impl2.trait_ref.instantiate_identity();
```

Comment:
```suggestion
        let trait_ref1 = impl1.trait_ref.skip_binder();
        let trait_ref2 = impl2.trait_ref.skip_binder();
```
existing, but the impls are in different environments, so we sohuld not use `instantiate_identity` here. (there's currently no difference between these functions for `EarlyBinder`, so this is only to declare intent)

Reply:
I feel it might be better to use `instantiate_identity` rather than `skip_binder` because
1.  we don't relate the two trait refs in this function,
2. `trait_ref.referenes_error()`  is called for each trait refs, which depends on generic parameters, so if we try to be strict on the `skip_binder` usage, this might cause a problem in the future when (e.g.,) `skip_binder` implements a protected substitution that forbids the access to the substituted types. 

Another feasible option might be to remove the `referenes_error` check here since a trait ref is already filtered out in the candidate assembly phase if it contains an error, at least in the current implementation. This change may make the function fragile slightly, though.

But I might misunderstand the semantics of `skip_binder`, I'd be helpful if you kindly correct me in that case.

Reply:
>> Another feasible option might be to remove the referenes_error check here since a trait ref is already filtered out in the candidate assembly phase if it contains an error

Hmm, but it'd be might be better to select a candidate eagerly for better diagnostics?
```rust
pub trait Foo<T> {
    fn foo(&self) {}
}
impl Foo<T> for i32 {}

fn foo() {
    let x = 1i32;
    x.foo();
}

fn main() {}
```
↓
```
error[E0412]: cannot find type `T` in this scope

  |
4 | impl Foo<T> for i32 {}
  |          ^ not found in this scope

error[E0599]: no method named `foo` found for type `i32` in the current scope
  |
8 |     x.foo();
  |       ^^^ method not found in `i32`
```
The second error message`[E0599]` might be redundant.

Reply:
The meaning of `fn EarlyBinder::skip_binder` should directly mirror `fn Binder::skip_binder`. After calling `skip_binder`, any `ty::Param/ty::Bound` are "escaping". Escaping bound vars are not supported in a lot of places, most notably, relating types and using the trait solver. 

> 1. we don't relate the two trait refs in this function,

Relating trait refs after `skip_binder` would be wrong, so if we relate them, we need to instantiate them, either with `instantiate_identity` or just `instantiate`.

> 2. `trait_ref.referenes_error()`  is called for each trait refs, which depends on generic parameters; this might cause a problem in the future when (e.g.,) `skip_binder` implements a protected substitution that forbids access to the substituted param types. 

I very much don't expect that to be the case. `trait_ref.references_errors` can currently be used for types even if they have escaping bound vars, and I expect that to stay this way.

> Another feasible option might be to remove the `referenes_error` check here since a trait ref is already filtered out in the candidate assembly phase if it contains an error, at least in the current implementation. This change may make the function fragile slightly, though.

Does removing the `references_error` check worsen diagnostics? I do think that we should actually do the reverse treat `ty::Error` as unifying with everything and not filtering impls

Reply:
Thanks for the clarification!

> Relating trait refs after skip_binder would be wrong

yeah, I didn't mean relating types after calling `skip_binder`. I mean, applying `instantiate_identity()` to both types would also be fine, even when the types come from different environments, as long as we don't relate them. 
But I understand that `skip_binder` is better since it gives stricter restrictions to forbid relating them. Thanks!

>  very much don't expect that to be the case.

Alright, it makes sense.

>I do think that we should actually do the reverse treat ty::Error as unifying with everything and not filtering impls.

Yes, I do agree with this, as I mentioned in "there might be a plan to collect candidates eagerly for better diagnostics"